x-pack/plugins/security/server/authentication/providers/saml.ts

```diff
@@ -234,22 +242,36 @@ export class SAMLAuthenticationProvider extends BaseAuthenticationProvider {
     this.logger.debug(`Trying to log user out via ${request.url.path}.`);
 
     // Normally when there is no active session in Kibana, `logout` method shouldn't do anything
-    // and user will eventually be redirected to the home page to log in. But when SAML is enabled
-    // there is a special case when logout is initiated by the IdP or another SP, then IdP will
-    // request _every_ SP associated with the current user session to do the logout. So if Kibana,
-    // without an active session, receives such request it shouldn't redirect user to the home page,
-    // but rather redirect back to IdP with correct logout response and only Elasticsearch knows how
-    // to do that.
-    const isIdPInitiatedSLO = isSAMLRequestQuery(request.query);
-    if (!state?.accessToken && !isIdPInitiatedSLO) {
+    // and user will eventually be redirected to the home page to log in. But when SAML SLO is
+    // supported there are two special cases that we need to handle even if there is no active
+    // Kibana session:
+    //
+    // 1. When IdP or another SP initiates logout, then IdP will request _every_ SP associated with
+    // the current user session to do the logout. So if Kibana receives such request it shouldn't
+    // redirect user to the home page, but rather redirect back to IdP with correct logout response
+    // and only Elasticsearch knows how to do that.
+    //
+    // 2. When Kibana initiates logout, then IdP may eventually respond with the logout response. So
+    // if Kibana receives such response it shouldn't redirect user to the home page, but rather
+    // redirect to the `loggedOut` URL instead.
+    const isIdPInitiatedSLORequest = isSAMLRequestQuery(request.query);
+    const isSPInitiatedSLOResponse = isSAMLResponseQuery(request.query);
+    if (!state?.accessToken && !isIdPInitiatedSLORequest && !isSPInitiatedSLOResponse) {
       this.logger.debug('There is no SAML session to invalidate.');
       return DeauthenticationResult.notHandled();
     }
 
     try {
-      const redirect = isIdPInitiatedSLO
+      // It may _theoretically_ (highly unlikely in practice though) happen that when user receives
+      // logout response they may already have a new SAML session (isSPInitiatedSLOResponse == true
+      // and state !== undefined). In this case case it'd be safer to trigger SP initiated logout
+      // for the new session as well.
+      const redirect = isIdPInitiatedSLORequest
         ? await this.performIdPInitiatedSingleLogout(request)
-        : await this.performUserInitiatedSingleLogout(state?.accessToken!, state?.refreshToken!);
+        : state
+        ? await this.performUserInitiatedSingleLogout(state.accessToken!, state.refreshToken!)
+        : // Once Elasticsearch can consume logout response we'll be sending it here.
```

Comment:
nit: can we link to https://github.com/elastic/elasticsearch/issues/40901 in this comment for future reference?
```suggestion
        : // Once Elasticsearch can consume logout response we'll be sending it here. See https://github.com/elastic/elasticsearch/issues/40901
```

Reply:
Yep, it's a good idea!